src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,421 @@
+    by simp [free_add_monoid.lift_eval_of, C_add]
+| _, _, (eqv.of_add_scalar z₁ z₂ f)      := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_add_scalar]
+| _, _, (eqv.of_smul z f i r')     := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_smul]
+| _, _, (eqv.add_comm x y)         := (add_con.ker_rel _).2 $
+    by simp_rw [add_monoid_hom.map_add, add_comm]
+end
+
+instance : has_scalar R (⨂[R] i, s i) :=
+⟨λ r, lift_add_hom (λ f : R × Π i, s i, tprod_coef R (r * f.1) f.2)
+  (λ r' f i hf, by simp_rw [zero_tprod_coef' _ f i hf])
+  (λ f, by simp [zero_tprod_coef])
+  (λ r' f i m₁ m₂, by simp [add_tprod_coef])
+  (λ r' r'' f, by simp [add_tprod_coef', mul_add])
+  (λ z f i r', by simp [smul_tprod_coef, (show r' * (r * z) = r * (r' * z), by ring)])⟩
+
+lemma smul_tprod_coef' (r z : R) (f : Π i, s i) :
+  r • (tprod_coef R z f) = tprod_coef R (r • z) f := rfl
+
+lemma smul_tprod (r : R) (f : Π i, s i) (i : ι) :
+  r • (tprod R f) = tprod R (update f i (r • f i)) :=
+by simp [tprod, smul_tprod_coef' r 1 f, smul_tprod_coef]
+
+lemma smul_tprod' (r : R) (f : Π i, s i) (i : ι) (m : s i) :
+  r • tprod R (update f i m) = tprod R (update f i (r • m)) :=
+begin
+  set g := update f i m with hg,
+  have h₁ : update f i (r • m) = update g i (r • m) := by simp_rw [hg, update_idem],
+  have h₂ : g i = m := by simp [hg],
+  rw [h₁, smul_tprod r _ i, h₂],
+end
+
+lemma tprod_coef_eq_smul_tprod (z : R) (f : Π i, s i) : tprod_coef R z f = z • tprod R f :=
+begin
+  have : z = z • (1 : R) := by rw [smul_eq_mul, mul_one],
+  conv_lhs { rw [this] },
+  rw [←smul_tprod_coef', ←tprod],
+end
+
+protected theorem smul_zero (r : R) : (r • 0 : ⨂[R] i, s i) = 0 :=
+add_monoid_hom.map_zero _
+
+protected theorem smul_add (r : R) (x y : ⨂[R] i, s i) :
+  r • (x + y) = r • x + r • y :=
+add_monoid_hom.map_add _ _ _
+
+@[elab_as_eliminator]
+protected theorem induction_on
+  {C : (⨂[R] i, s i) → Prop}
+  (z : ⨂[R] i, s i)
+  (C1 : ∀ {r : R} {f : Π i, s i}, C (r • (tprod R f)))
+  (Cp : ∀ {x y}, C x → C y → C (x + y)) : C z :=
+begin
+  have C0 : C 0,
+  { have h₁ := @C1 0 0,
+    rwa [tprod, smul_tprod_coef', zero_smul, zero_tprod_coef] at h₁ },
+  refine add_con.induction_on z (λ x, free_add_monoid.rec_on x C0 _),
+  simp_rw add_con.coe_add,
+  refine λ f y ih, Cp _ ih,
+  simp_rw [tprod, smul_tprod_coef', smul_eq_mul, mul_one] at C1,
+  convert @C1 f.1 f.2,
+  simp only [prod.mk.eta],
+end
+
+@[elab_as_eliminator]
+protected theorem induction_on'
+  {C : (⨂[R] i, s i) → Prop}
+  (z : ⨂[R] i, s i)
+  (C1 : ∀ {r : R} {f : Π i, s i}, C (tprod_coef R r f))
+  (Cp : ∀ {x y}, C x → C y → C (x + y)) : C z :=
+begin
+  have C0 : C 0,
+  { have h₁ := @C1 0 0,
+    rwa [zero_tprod_coef] at h₁ },
+  refine add_con.induction_on z (λ x, free_add_monoid.rec_on x C0 _),
+  simp_rw add_con.coe_add,
+  refine λ f y ih, Cp _ ih,
+  convert @C1 f.1 f.2,
+  simp only [prod.mk.eta],
+end
+
+instance : semimodule R (⨂[R] i, s i) :=
```

Comment:
Note that in #5317 I change the definition of `semimodule` for `tensor_product` - it might make sense to wait for that to be merged or rejected before merging this.

Reply:
I don't think it would affect this PR directly -- I mean it would be nice for both this and the binary tensor product to be constructed the same way, but it won't cause any merge conflicts either way.

Reply:
That PR has now been merged, so it would be nice to get the `is_scalar_tower`-based instance in here too

Reply:
Done!